Sources/NIO/EventLoopFuture.swift

```diff
@@ -1044,7 +1054,76 @@ extension EventLoopFuture {
     ///     - on: The `EventLoop` on which the new `EventLoopFuture` callbacks will fire.
     /// - Returns: A new `EventLoopFuture` with all of the values fulfilled by the provided futures.
     public static func whenAllSucceed(_ futures: [EventLoopFuture<Value>], on eventLoop: EventLoop) -> EventLoopFuture<[Value]> {
-        return .reduce(into: [], futures, on: eventLoop) { (results, value) in results.append(value) }
+        let promise = eventLoop.makePromise(of: Void.self)
+
+        var results: [Value?] = .init(repeating: nil, count: futures.count)
+        let callback = { (index: Int, result: Value) in
+            results[index] = result
+        }
+
+        if eventLoop.inEventLoop {
+            self._reduceSuccesses0(promise, futures, eventLoop, onValue: callback)
+        } else {
+            eventLoop.execute {
+                self._reduceSuccesses0(promise, futures, eventLoop, onValue: callback)
+            }
+        }
+
+        return promise.futureResult.map {
+            // verify that all operations have been completed
+            assert(!results.contains(where: { $0 == nil }))
+            return results.map { $0! }
+        }
+    }
+
+    /// Loops through the futures array and attaches callbacks to execute `onValue` on the provided `EventLoop` when
+    /// they succeed. The `onValue` will receive the index of the future that fulfilled the provided `Result`.
+    ///
+    /// Once all the futures have succeed, the provided promise will succeed.
+    /// Once any future fails, the provided promise will fail.
+    private static func _reduceSuccesses0<InputValue>(_ promise: EventLoopPromise<Void>,
+                                                      _ futures: [EventLoopFuture<InputValue>],
+                                                      _ eventLoop: EventLoop,
+                                                      onValue: @escaping (Int, InputValue) -> Void) {
+        eventLoop.assertInEventLoop()
+
+        var remainingCount = futures.count
+
+        if remainingCount == 0 {
+            promise.succeed(())
+            return
+        }
+
+        // Sends the result to `onValue` in case of success and succeeds/fails the input promise, if appropriate.
+        func processResult(_ index: Int, _ result: Result<InputValue, Error>) {
+            switch result {
+            case .success(let result):
+                onValue(index, result)
+                remainingCount -= 1
+
+                if remainingCount == 0 {
+                    promise.succeed(())
+                }
+            case .failure(let error):
+                promise.fail(error)
+            }
+        }
+        // loop through the futures to chain callbacks to execute on the initiating event loop and grab their index
+        // in the "futures" to pass their result to the caller
+        for (index, future) in futures.enumerated() {
+            if future.eventLoop === eventLoop,
```

Comment:
I’m not sure about the safety of this operation. Why are we checking for loop equality rather than asking if we’re on that future’s event loop?

Reply:
Good point; I think this was mostly left over from a previous implementation and the uninformed belief that `inEventLoop` could be slower. Changed to use `inEventLoop` now, performance is essentially the same:

```
measuring: future_whenallsucceed_100k_immediately_succeeded_off_loop: 0.13757002353668213, 0.13881397247314453, 0.13828301429748535, 0.13999199867248535, 0.1392430067062378, 0.1379610300064087, 0.14011895656585693, 0.13859009742736816, 0.13950490951538086, 0.13902902603149414, 
measuring: future_whenallsucceed_100k_immediately_succeeded_on_loop: 0.14021098613739014, 0.13878905773162842, 0.1398860216140747, 0.13852500915527344, 0.1395789384841919, 0.13947105407714844, 0.1404019594192505, 0.13997399806976318, 0.14212501049041748, 0.138746976852417, 
measuring: future_whenallsucceed_100k_deferred_off_loop: 0.8842309713363647, 0.8726869821548462, 0.8777910470962524, 0.880282998085022, 0.8623709678649902, 0.8720219135284424, 0.8791440725326538, 0.8786169290542603, 0.8717180490493774, 0.881566047668457, 
measuring: future_whenallsucceed_100k_deferred_on_loop: 0.20134305953979492, 0.1984039545059204, 0.19959497451782227, 0.20501303672790527, 0.20239198207855225, 0.20087504386901855, 0.2013930082321167, 0.2047710418701172, 0.2020859718322754, 0.19875800609588623, 
measuring: future_whenallcomplete_100k_immediately_succeeded_off_loop: 0.11057305335998535, 0.11202895641326904, 0.11022496223449707, 0.11104798316955566, 0.10982406139373779, 0.10918903350830078, 0.10978901386260986, 0.11864101886749268, 0.11072397232055664, 0.10935401916503906, 
measuring: future_whenallcomplete_100k_immediately_succeeded_on_loop: 0.10864400863647461, 0.1091160774230957, 0.10998594760894775, 0.11082005500793457, 0.10957396030426025, 0.1101609468460083, 0.1113889217376709, 0.10998702049255371, 0.10958206653594971, 0.10873699188232422, 
measuring: future_whenallcomplete_100k_deferred_off_loop: 0.8409409523010254, 0.8487280607223511, 0.8519699573516846, 0.8609169721603394, 0.857483983039856, 0.8517990112304688, 0.8563249111175537, 0.8470900058746338, 0.8306020498275757, 0.8549350500106812, 
measuring: future_whenallcomplete_100k_deferred_on_loop: 0.17533302307128906, 0.1744459867477417, 0.17379295825958252, 0.17324697971343994, 0.17416608333587646, 0.17526590824127197, 0.17662596702575684, 0.17753005027770996, 0.1744699478149414, 0.1766829490661621, 
```